### PR TITLE
Reorganize build system and make host library configurable

### LINKED ISSUE #36: [RFC] Reorganize build system
I want to use lkl in one of my projects and currently the bulid process is a bit clumsy. There ist no `make install` and the build is only within the tree, and the Kernel's Kbulid won't install lkl headers like `lkl.h` and similar issues that are all minor by themselves.

I therefore want to volunteer to clean up the build system a little bit, provide a sensible install target, provide a .pc file for `pkg-config` and maybe make the featureset of lkl more configurable (e.g. separate the liblkl* stuff from current userspace tools like `lklfuse` and `fs2tar`. 

Before I start, I have two questions:
1. Is this a welcomed thing at all? Or is lkl not intended to be used like e.g. libm or libpthread but only for people who definitely know what they are doing (and do it manually)?
2. Is GNU autotools an acceptable build dependency? I'm most familiar with autotools and I don't feel Kbuild is used too much in tools/ (I think to add config options, these would have to go into arch/lkl?). The only other user of autoconf is tools/usb/usbip, others just use environment variables or Makefile variables that have to be edited.

### PR DESCRIPTION
**Pull request by tommie-lie**
Finally the patch series I promised to provide in #36.
This pull request introduces a new build system that is also hooked into the main kernel build system. You can now configure lkl from the kernel tree's root directory using any of the *config make targets. The host library and sample tools have their own menu (at the bottom of the configuration root).

This paves the way to have configurable features in the host library (like different network backends).

The default config still compiles all the libraries and all tools but you can enable and disable them individually (e.g. compiling just the static library instead of shared objects).

There is also a commit that adds a target to install the host libraries and header to appropriate directories. This makes it easier for third-party applications to make use of lkl.

It don't really mind if this is merged after #67. I can provide patches to either include the DPDK stuff here or use the new buildsystem in the DPDK branch. However, this PR includes some fixes that were already merged to the old Makefile (and I think it also fixes the compile issue from #80) and merging this PR sooner rather than later will prevent people from doing unnecessary double work.

(Fixes #36)

[Review on Reviewable](https://reviewable.io/reviews/lkl/linux/82)

### CODE REVIEW

**Comment by opurdila** [comment]
Looks good to me. I will wait for a couple of days before merging to get feedback from the other developers.

**Comment by thehajime** [comment]
Reviewed 1 of 13 files at r1.
Review status: 1 of 13 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/lkl/linux/82#-KB5GzqvEPVEBn4UDClq:-KB5GzqvEPVEBn4UDClr:-2052652820) (no related file):*
I think all are fine ! put only a minor comment.

---

*[tools/lkl/Kconfig, line 27 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/82#-KB5FfezFBuylv13DcE4:-KB5Fff-5oMq5lb5TJcZ:-2011036376) ([raw file](https://github.com/lkl/linux/blob/7703a797a38cb6ecda92e20399d14a910c18ac41/tools/lkl/Kconfig#L27)):*
I think ```bool "lkl hijack library"``` is  appropriate (NUSE only expresses about networking code but hijack is not limited to)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82)*

**Comment by thehajime** [comment]
Reviewed 12 of 13 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82)*

**Comment by pscollins** [comment]
Looks great! I'm not familiar with the convention of naming files `Build`? Are those normal `Makefile`s, or something else?

---

Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82#-:-KB5vzLJOMTwIsvJcGYg:-1464431603)*

**Comment by tommie-lie** [comment]
Those are *kind of* normal Makefiles and belong to the tools/build framework. You can think of them as serving the same purpose as Kbuild files used throughout the Kernel tree. They define which files to compile/link and are (ultimately) included from tools/build/Makefile.include so they are Makefile syntax. See tools/build/Documentation for more info.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tools/lkl/Kconfig, line 27 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/82#-KB5FfezFBuylv13DcE4:-KB7WDXTRMOwU6TEev7x:2126977310) ([raw file](https://github.com/lkl/linux/blob/7703a797a38cb6ecda92e20399d14a910c18ac41/tools/lkl/Kconfig#L27)):*
I did not think about it when I wrote this but now that you mention it: I think NUSE is still appropriate at the current state of the hijack library. (Aside from close, read and write) it only hijacks network operations. Without open and more infrastructure to properly support LKL's file system support, liblkl-hijack is little more than NUSE, I'm afraid.

But I'm really not clinging to that config label. If the others agree to you, I'll happily change it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82#-:-KB7UuzEKgLXInpXfVv5:-1493489038)*

**Comment by pscollins** [comment]
Ah, so it's a part of the kernel build system I didn't know about. Thanks!

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82#-:-KB7_JLrHBudshlgD5Rk:1883213708)*

**Comment by pscollins** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[tools/lkl/Kconfig, line 27 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/82#-KB5FfezFBuylv13DcE4:-KBAVFzPz3K15grALvp7:-82679874) ([raw file](https://github.com/lkl/linux/blob/7703a797a38cb6ecda92e20399d14a910c18ac41/tools/lkl/Kconfig#L27)):*
Either way is fine, in my opinion. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82)*

File: tools/lkl/Makefile [hunk 1]
```diff
@@ -1,95 +1,205 @@
...
+lklfuse$(EXESUF): lklfuse-in.o liblkl.a
+lklfuse$(EXESUF): LDLIBS += -lfuse
 
-ifeq ($(OUTPUT_FORMAT),elf64-x86-64)
-lib += lib/liblkl-hijack.so
-hijack_objs = $(patsubst %.c,%.o,$(wildcard lib/hijack/*.c))
-objs += $(hijack_objs)
+fs2tar$(EXESUF): fs2tar-in.o liblkl.a
+fs2tar$(EXESUF): LDLIBS += -larchive
+ifneq (,$(filter $(OUTPUT_FORMAT),elf64-x86-64-freebsd))
```
Comment (stfairy) [inline tools/lkl/Makefile]: Why freebsd matters here?

**Comment by stfairy** [comment]
LGTM.  This looks much nicer!

**Comment by opurdila** [comment]
I would like to avoid pulling tools/lkl dependencies on the kernel build and Kconfig / .config. So far none of the tools do this and I think we will have a hard time justifying it.

What would we lose if we remove d6e675f ?

**Comment by tommie-lie** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tools/lkl/Makefile, line 115 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/82#-KB3INbrMNsg69Gr8VK0-r1-115:-KBC_5h_naVXeutE1gdK:-1405165341) ([raw file](https://github.com/lkl/linux/blob/7703a797a38cb6ecda92e20399d14a910c18ac41/tools/lkl/Makefile#L115)):*
See the next line. (Free)BSD has libargp while GNU/Linux has argp support builtin in the libc. The old Makefile has a similar conditional: https://github.com/lkl/linux/blob/master/tools/lkl/Makefile#L87

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82)*

**Comment by tommie-lie** [comment]
@opurdila 
tl;dr: Having LKL configurable is a **very** nice thing to have if you don't want/need all the subsystems. Currently, the benefits are limited (or even negligible) but I already know of two applications for it (DPDK and VDE). I think the workings on liblkl already depend on how the Kernel is built (in order not to break liblkl) so liblkl should also know about it.

Right now, the Kconfig stuff is a little pointless, building the host libraries and tools does not really hurt. But together with #67 and my PR for VDE (in the pipeline), it makes more sense to have a configurable host library: Select if you want tuntap, DPDK or VDE networking support (or any combination thereof) only if VIRTIO_NET is also enabled. Remove networking or filesystem stuff completely from the build: If all you're interested in is NUSE-like features, you can reduce the build size by removing all the filesystems in which case you also don't need the virtio-blk stuff from the host library. You see, there are a lot "if this subsystem, then that feature" conditions that can be thought of. Kconfig has the advantage that it does all the sanity checking for us and provides a nice interface. It also checks the input arguments, so it does not matter if you pass (in terms of #67) "dpdk=yes", "dpdk=y" or "dpdk=enable" to make, Kconfig will only allow the valid variant.

You are right in that it is the first tool to actually make use of include/config/auto.conf but I would justify it by saying all the other tools don't rely that heavily on how the kernel is built (I think LKL will just break right now if the kernel is built without virtio). LKL even includes an object built from the main kernel tree (lkl.o), so it has to know what's going on.
I got the original idea from http://thread.gmane.org/gmane.linux.kernel/1792367 which aimed at building tools/perf with Kbuild/Kconfig. It was ultimately replaced by tools/build and no config, but I was not able to find a justification against the concept of using Kconfig. Alexis Berlemont's patch also modified Kconfig so that it was possible to use Kconfig **outside** the kernel tree and only configure tools/perf, which is really neat. I did not want to be as intrusive as that given that tools/build exists now. The ugliest thing about my PR is that arch/lkl/Kconfig includes tools/lkl/Kconfig and I did not find a nicer way to do that.
The alternative would be autoconf which many think of as even worse and if arch/lkl is to be merged, tools/lkl will become a core component to it.

On a related matter: I think introducing tools/build was already a mistake and Alexis Berlemont's approach would have been more future-proof and would have completely eliminated this discussion. But that is neither for me to decide nor can we change something about it right now. However, I think re-introducing his changes to Kbuild and Kconfig in the lkl tree is out of the question.

**Comment by thehajime** [comment]
while I like @tommie-lie 's PR to configure lkl tool's related options via ```make menuconfig```, I also like the approach of @opurdila to be _quiet_ for a while.

I would like to recall @opurdila 's question since I may miss the answer from your comment.

> What would we lose if we remove d6e675f ?

**Comment by tommie-lie** [comment]
On 23.02.2016 14:29, Hajime Tazaki wrote:
> I would like to recall @opurdila 's question since I may miss the answer from your comment.
> 
>> What would we lose if we remove d6e675f ?

Sorry, I hoped that was clear: In this PR (the 4 commits at hand), not
much and nothing too important is lost by removing d6e675f.
Currently, you can select which flavor of the host library (static,
shared) you want to build and which of the tools (hijack library,
lklfuse, cptofs and fs2tar). Additionally, `lib/lkl.o` will be
recompiled if the configuration changes, reflecting the changes in
config without explicitly having to clean/remove `lib/lkl.o`.

In the old Makefile, `lib/lkl.o` is never rebuilt, even if it has to, so
removing d6e675f does not introduce a regression. Without the dependency
to the Kernel's config, it is impossible to determine if `lib/lkl.o`
should be rebuilt, anyway.

Not merging d6e675f from this PR would just always build all libraries
and all binaries, i.e. completely retain the status quo.

**Comment by thehajime** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tools/lkl/Kconfig, line 27 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/82#-KB5FfezFBuylv13DcE4:-KBHinzA3ux-jwqAy6BL:-1451084267) ([raw file](https://github.com/lkl/linux/blob/7703a797a38cb6ecda92e20399d14a910c18ac41/tools/lkl/Kconfig#L27)):*
for the current state, yes, but it wasn't meant to introduce only for nuse in my motivation so, it would be nice to be _without_ NUSE in the label.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82)*

**Comment by thehajime** [comment]
thanks for the clarification. indeed, building always lib/lkl.o is annoying..

I'm still neutral between @opurdila and @tommie-lie .

**Comment by tommie-lie** [comment]
I just changed the Kconfig commit as per @thehajime 's comment on reviewable.io (remove the reference to NUSE for CONFIG_LKL_HIJACK). Please note the changed hashes for the last two commits in this series.

**Comment by thehajime** [comment]
@tommie-lie thanks.

**Comment by thehajime** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/82)*

**Comment by opurdila** [comment]
@tommie-lie I will skip the Kconfig stuff for now - we can revisit that part when we have the need to configure the host libs. 

I've pushed your patches (+ a minor hijack test fix - squashed in the original commit) to build_test branch and I will merge them to master soon.

Thanks a lot Thomas for this, the build system looks much better now :)

**Comment by tommie-lie** [comment]
@opurdila Thanks. I will come back with a new Kconfig patch when it provides more features ;-)
